Add `Anytime` and `Until`

lucianojoublanc-da opened the pull request:
This adds Anytime and Until nodes. We've avoided them so far as the semantics are trickier, but I've included them here in favour of adding complexity to Observation #13 . cc @leonidr-da

There are still some bits that are unimplemented - you'll get warnings when you build - notably, the lifecycle operation. I'm currently working on fixing #10 which will change that operation, so would hold off refactoring lifecycle until that's been resolved (the refactor should make it easier to work with it, hopefully (!).

lucianojoublanc-da commented:
I've realised that we need to think about adding a function to prune expired branches. This wasn't really necessary so far with the way we implemented `lifecycle` - all the primitives expired at acquisition time, I think. However, at least for `Until` this is not the case.

This will probably be done by having a new `pruneExpired : t -> Claim -> Claim`, and also a `lifecycleAndExpire` that does everything in a single step.

leonidr-da commented:
In general banking terminology, to "lifecycle" is also to get rid of things that have expired.

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
 
--- | Returned from a lifecycle operation.
+-- | Returned from a `settle` operation.
 data Result a = Result with
```
Rename to `SettleResult`

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
-  . apoCataM (pruneZeros') (lifecycle' spot choose)
+  . apoCataM pruneZeros' settleLazily'
+    where
+      settleLazily' = (lift . lift <$> lazily' spot)
+                      >=> (fmap distM (fmap (settle' spot . embed)) . sequence)
```
This is where all the complexity is shoved away to. All this does is to compose `lazily'` and `settle'` and sequence the effects to have the right type signature: `C a -> M (F a (C a))`.

lucianojoublanc-da replied:
Consider writing a combinator that does same as `apoCataM` but you can pass a `C a -> M (F a (C a))` as the second argument, and `lazily` is already pre-applied.

lucianojoublanc-da replied:
Not quite what I wrote above, but I did implement a general `ghyloM`.

lucianojoublanc-da commented:
I've committed a preview of the refactoring I've done on lifecycling.

Functionality
----------------

I've split the old `lifecycle` function into three new API functions, following discussion with @leonidr-da .
- `settle`, which does the same as the old function, but aborts on `Or` nodes now. i.e. users must first get rid of those nodes using the new function ...
- `exercise`, which takes a tree as an argument, and replaces any matching `Or`'s with this tree. This is also used to resolve an `Anytime` node (i.e. `exercise (Anytime c) -> c`.
- `expire`. This is used to prune any branches that have expired, particularly `Until`.

This still leaves it up to the implementer to use these correctly. E.g., `exercise` should only be used by the bearer, and needs to be run _before_ `settle`.

With this new set-up we should now be able to handle option exercises much better, and also support `Anytime` and `Until` fully.

Implementation
--------------------

I've split the old `lifecycle'` function into a number of different functions, to make it easier to reason about the different effects:

- `lazily'`. This doesn't transform the tree in any way; it needs to be pre-composed with other functions, to 'short-circuit' the traversal when a condition is false. Now that this logic is no longer in the other functions, you **must** remember to first apply this function; otherwise e.g. running `settle'` on it's own will settle _all_ nodes in the tree, not only the current ones.
- `settle'` which has the `Writer` monad for logging `One` leaves that need to be settled.
- `exercise'` which replaces `Or` and `Anytime`s with the passed-in branches.

There are a couple more. But the point now is that each of these functions is succinct, easy to reason about, and this means that we can test these separate parts independently.

Performance wise, the traversals are still O(n), as all the functions are fused together in calls to `apoParaM`. There's probably some overhead for the sequencing steps, but I'm hoping this is minimal.

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
 module ContingentClaims.Lifecycle (
-    lifecycle
-  , lifecycle'
+    settle
```
export `exercise` and `expire`.

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
-lifecycle' spot choose (Cond obs c c') = do
-  predicate <- lift . lift $ eval spot obs
-  if predicate then lifecycle' spot choose c else lifecycle' spot choose c'
+  modify (* k)
```
See #18 ; I think this can be fixed by reverting to a `Reader` monad.

lucianojoublanc-da replied:
Was fixed by embedding the state in the carrier, i.e. `a` in `a -> f a`.

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
+settle' _ other = pure $ project other
+
+-- | Used to make an election on `Or` or exercise an `Anytime`.
+--  `import` this `qualified`, to avoid clashes with `Prelude.exercise`.
```
I'm aware of the name clash, but this is unfortunately the best term available. I considered `elect` also.

I don't see any problem using it qualified e.g. `Lifecycle.exercise`.

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
+distM (Right m) = fmap Right <$> m
+
+-- | Evaluate observables and skip branches for which predicates don't hold.
+-- This is useless onl it's own. Composed with other functions it adds laziness.
```
typo

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
--- | Returned from a lifecycle operation.
+-- | Returned from a `settle` operation.
 data Result a = Result with
   remaining : C a
```
Consider swapping order of arguments, for consistency with other data types. The functor tends to be on the right-hand-side. (Think how you'd repeatedly fold over this result - on the `remainder`).

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
+  pure if predicate then WhenF obs (Right c) else WhenF obs (Left c)
+lazily' spot (Cond obs c c') = do
+  predicate <- eval spot obs
+  if predicate then lazily' spot c else lazily' spot c'
```
Remove these `lazily'`, as @leonidr-da pointed out, explicit recursion is confusing here.

lucianojoublanc-da replied:
This can't be done without changing this to a futumorphism. That adds yet another type (`Free`) that is used to 'unfold' the tree several levels ahead. In this case, it's safe to call the op recursively, as we are only 'consuming' new nodes. I'd like to change this, but I'm afraid that doing so would add even more complexity at the type-level, so I'll leave as-is for now.

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
+lazily' spot (Cond obs c c') = do
+  predicate <- eval spot obs
+  if predicate then lazily' spot c else lazily' spot c'
+lazily' _ other = pure $ Right <$> project other
```
> leonidr  13 hours ago
> Thinking about the full lifecycle a bit more deeply, I realized that we can't necessarily lifecycle Give automatically either. It is much more than multiplying the payments inside by -1. Imagine if there's an Or inside that Give. What we've done is changed who gets decide how to proceed with that portion of a claim. We can have the "limited" view of Give where it is effectively Scale -1 but with a different name. But I think that within a Daml framework, the "intended" interpretation of Give would require changing controller rights (assuming both the counterparties are signatories).

That's a great point @leonidr-da . The approach I was considering was to also `short-circuit` on a `Give` node here (or possibly in a **separate algebra**). That means that, composed with `exercise`, you would never be able to exercise an `Or` or `Anytime` below a `Give` node.
I'd then need a way to flip that around (so for example a callable bond gives the issuer the right - not the bearer).
In either case, splitting up the effects into separate functions keeps the handling of the `Scale` factor and the concept of who can exercise entirely separate (and can be tested independently).

lucianojoublanc-da replied:
One interesting thing to consider is whether `Give` should only restrict exercise rights, or also settlement rights - i.e. should `One`'s below `Give` only be settled by the counterparty, or by either party?

This is a liveness problem.

One discussion I had with @GeorgSchneider was whether the custodian (counterparty) should be responsible for _initiating_ settlement in call cases. For instance, in the real-world, you don't need to do anything to receive coupons or dividends in your brokerage account; that's part of the safe-keeping role of the custodian. However, not sure if this changes in a trustless environment. As we're assuming here that all claims are positive, then would it make sense that each counterpart should have the right to trigger settlement of their own claims? But then this has the added inconvenience of requiring the bearer to settle his own dividends!
To further complicate matters, recall that the `settle` (previously `lifecycle`) doesn't actually _do_ the settlement - it merely indicates _what needs to be settled_.

leonidr-da replied:
> be able to exercise an Or or Anytime below a Give node.

I think it is worse than that. If during the lifecycling of a claim, in the broadest interpretation of `Give`, in a trustless environment, I do not have the _right_ to even lifecycle the claim inside. What if it contains an `Observation` to something that at the moment is off-ledger (to be put on ledger) that the new holder is privy to?

leonidr-da replied:
I think we need to differentiate lifecycling based on who is performing it. In a custodian/clearing-house case (with delegation from the parties), we can have explicit instructions. Such a party would also have restrictions on what time of assets it is willing to lifecycle.

lucianojoublanc-da replied:
Here's another mind-bender: nested `Give`s. How would you handle this:
```haskell
Give (One "USD" `And` Give (One "EUR"))
```

lucianojoublanc-da replied:
The most recent commit adds a boolean argument to the election parameter of `Lifecycle.exercise`, which indicates who is making the election (`True` = the bearer, resp. counterparty).

https://github.com/digital-asset/contingent-claims/pull/15/files#diff-777017d2d9038baba5d5232cbf1a037d054a2d5b65acec36099bea4385efed71R118

It handles nested cases correctly.

For other operations, the authorization for `Give` is ignored; any party can initiate settlement when it's due, or expire a contract past it's horizon (maturity). These help to ensure liveness; i.e. one party might maliciously to 'do nothing' to block the claim from advancing.

leonidr-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,120 @@
+
+-- | Coalgebra for `exercise`.
+exercise' : Eq a => C a -> C a -> F a (C a)
+exercise' chosen (Or c _) | chosen == c = project c
```
I think that one should not be allowed to exercise on `Or` | `Anytime` unless it is a root node. You may not own them.

lucianojoublanc-da replied:
Here's a counter-example. How would you lifecycle this scenario:
```haskell
Scale 100.0 (c Or c')
```

lucianojoublanc-da replied:
> You may not own them.

I've not got as far as to handle `Give` nodes, but the idea is that that is handled by a separate function, that you will compose this with. I.e. this function will only ever receive branches where it is the 'owner'.

leonidr-da replied:
```
Scale 100.0 (c Or c') = (Scale 100.0 c) Or (Scale 100.0 c')
```

When we lifecycle a `Scale` we should move it to a terminal (next to a payment) position.

leonidr-da replied:
> but the idea is that that is handled by a separate function..

I think this is where no one solution will handle all situations. For true trustless defi one should stop evaluation. For a custodian case, the custodian should mandate instructions from the client and a default action otherwise.

Performing the recursion explicitly allows us to control evaluation without the need for a special `Reader`

lucianojoublanc-da replied:
See my previous comment. I believe we should be able to cover most situations now, as I have split the lifecycle and exercise into separate functions.
The `Lifecycle.exercise` now checks that only the bearer can make elections, accounting for `Give` nodes (even in nested expressions). The settlement function can be run by either party. These changes mean that we support any shape of `Claim`.

leonidr-da commented:
I like the enthusiasm of this approach but I am still a bit confused. In my mind the intent of a lifecycle function is to advance the claim; to convert the claim into payments/obligations and a resulting claim. You've previously identified the two inputs into this process

1. The state of the world. A way to reify observations.
2. Potential exercise instructions.

I am not certain why we need to decompose lifecycle into three methods. Your choice of names (ex. settle, expire) is also problematic as they already have overloaded meanings. Trades settle, not the instrument within a trade. (Most) Bonds do not expire they mature.

Lastly, I think that lifecycling should be lazy by default. We should not traverse the whole tree unless the claim compels us to.

lucianojoublanc-da commented:
> I am not certain why we need to decompose lifecycle into three methods. 

After reading some of your feedback, I think I'm beginning to come around to your way of thinking about this.
If I get you correctly, the two parties will take turns to lifecycle the instrument, peeling off layers of the `Claim` until they can no longer proceed. For example, until a (false) `When`, `Give` or a `Or` node is reached. Then the other party takes up the lifecycling. Is this correct?

> Your choice of names (ex. settle, expire) is also problematic as they already have overloaded meanings. Trades settle, not the instrument within a trade. (Most) Bonds do not expire they mature.

Noted. I'll think about it some more. Maybe keep `lifecycle` for `settle`. And perhaps `elect` for `exercise`.

> Lastly, I think that lifecycling should be lazy by default. We should not traverse the whole tree unless the claim compels us to.

I agree.

leonidr-da commented:
> Then the other party takes up the lifecycling. Is this correct?

As I said in another part: I think this is where no one solution will handle all situations. For true trustless defi one should stop evaluation. For a custodian case, the custodian should mandate instructions from the client (s) and a default action otherwise.

lucianojoublanc-da commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -3,83 +3,119 @@
+-- This function **ignores** conditions; it traverses the entire tree.
+settle' : CanAbort m 
+        => (Text -> Date -> m Decimal) 
+        -> (Decimal, C a) -> WriterT [(Decimal, a)] (Kleisli m Date) (F a (Decimal, C a))
```
Had to get rid of the 'quantity' state monad and put the carrier in the input argument. The reason is that with `StateT`, the quantity was being accumulated _across_ branches. It's like we need an `Applicative` `StateT` in order to not accumulate across-branches. This is the next best thing. This fixes #18 .

lucianojoublanc-da commented:
@chrisnorris-da this is now in a state which could be merged. There's still stuff missing, notably:
- Implement `LIfecycle.expire` function. This handles e.g. American options that haven't been exercised. It's not crucial as you won't be able to exercise them past expiry, but it's untidy as you can't replace the claims with a `Zero`.
- Handling of `Give` node. The idea will be that the `Lifecycle.exercise` will also return a value (probably a boolean) indicating who exercised the choice. It's then up to the implementer to check authorization. Currently an option can be `Lifecycle.exercise`d, but there is no way of knowing who was the 'owner' of the `Or`/`Anytime` node that was exercised.
- I'm thinking of swapping around the arguments of `Lifecycle.Result`, just so you can fold over the resulting tree.
- Thinking of swapping the arguments of `Until` constructor, so you can write e.g. 
```
Anytime (after t) _ `Until` maturity
```
- And some general cleanup

lucianojoublanc-da commented on test/daml/Test/Lifecycle.daml:
```diff
@@ -25,67 +31,230 @@ type C = Claim Observation Date Decimal Text
+  remaining === Until (O.TimeGte afterTomorrow) (Anytime (O.TimeGte tomorrow) (payoff `Or` Zero))
+  pending === []
+  
+  remaining <- Lifecycle.exercise observe25 (payoff `Or` Zero) remaining t
+  remaining === Until (O.TimeGte afterTomorrow) (payoff `Or` Zero)
+
+  remaining <- Lifecycle.exercise observe25 payoff remaining t
+  remaining === Until (O.TimeGte afterTomorrow) payoff
```
It's a bit annoying here that we need to lifecycle twice, once to remove the `Anytime` wrapper and a second time to remove the not-selected `Or` branch.
On the other hand, if say I would instead take as an argument the 'desired end state' (i.e. `payoff` in this case), the client would need to implement complicated logic to derive this; in the current form, they just take one of the child nodes of `Anytime` or `Or`; that's trivial.

GeorgSchneider commented on daml/ContingentClaims/FinancialClaim.daml:
```diff
@@ -48,17 +48,24 @@ fixed principal coupon = floating (O.pure principal) (O.pure coupon)
+american t maturity payoff = 
+  Until (at (succ maturity)) $
+    Anytime ((>=) t) (payoff `Or` Zero)
+  where (>=) = at
```
I don't get why you set `(>=)` to `at` here. `at` does date equality, so it reads now `anytime (today == t)`..?

lucianojoublanc-da replied:
`at` has been changed; if you look at the definition it's `TimeGte` (time greater than or equal).

The thing is that when using it when `When`, requires you to take _the first instant_ when the value is true. So effectively, at behaves like equality.
`Anytime` differs from `When`, because it allows an acquisition region. So for `Anytime`, having an equality is not desireable; otherwise it's just the same as `When`.

I recognise this is confusing though. It was done to reduce the ADT and number of cases that we have to handle.

GeorgSchneider commented on daml/ContingentClaims/Lifecycle.daml:
```diff
@@ -17,71 +21,152 @@ import ContingentClaims.Observation (Observation)
+  . apoCataM pruneZeros' acquireThenSettle
+  . (1.0, )
+  where
+    acquireThenSettle =
+      fmap (fmap (fmap join))
+      . fmap (fmap distE)
+      . fmap join
+      . fmap (fmap sequence)
+      . fmap (fmap (fmap (lifecycle' spot)))
+      . fmap (fmap sequence)
+      . fmap sequence
+      . fmap
+      . fmap (fmap (fmap embed))
+      . fmap (fmap (tsidE : F a (Either (C a) (C a)) -> Either (C a) (F a (C a)) ))
+      . fmap lift
+      $ acquire' spot
```
You'll have to talk me through this function :D

lucianojoublanc-da replied:
The composition of functions in `Lifecycle` can be simplified, using identities like `mapA f = sequence . map f`, `fmap f . fmap g = fmap (f . g)` etc. That will also improve performance. I'm raising a ticket for this.

GeorgSchneider commented on daml/ContingentClaims/Util.daml:
```diff
@@ -27,11 +31,21 @@ enum' (i, Cofree _ (GiveF f)) = CofreeF i (GiveF (succ i, f))
+--TODO should fail if dates will never be executed
+fixings' : ClaimF f t x a [t] -> [t]
+fixings' (WhenF (O.TimeGte t) ts) = t :: ts
+fixings' claim = fold claim
```
If I understand this function correctly, it pulls out dates at which things happen. Shouldn't the match then be on time equality, ie. `O.TimeEqu`, instead of `O.TimeGte`?

lucianojoublanc-da replied:
There's no time equality; as explained in my previous comment, the semantics of `<=` vary for `When`, which is a point operator, and `Anytime` which is a region operator.

I suppose we could add a `TimeEqu`, but it wouldn't behave any differently from `TimeGte`, and we'd need to write an extra case for any function we implement on `Claim`. What do you think @georg-da ? If you're confused by this, so will other people be too.

GeorgSchneider reviewed (approved):
Left some comments / questions to discuss. Otherwise, looks great!
